pkg/utils/tsoutil/tso_dispatcher.go

```diff
@@ -65,75 +74,62 @@ func (s *TSODispatcher) DispatchRequest(
 	ctx context.Context,
 	req Request,
 	tsoProtoFactory ProtoFactory,
-	doneCh <-chan struct{},
-	errCh chan<- error,
-	tsoPrimaryWatchers ...*etcdutil.LoopWatcher) {
+	tsoPrimaryWatchers ...*etcdutil.LoopWatcher) context.Context {
 	key := req.getForwardedHost()
 	val, loaded := s.dispatchChs.Load(key)
 	if !loaded {
-		val = make(chan Request, maxMergeRequests)
+		val = tsoRequestProxyQueue{}
 		val, loaded = s.dispatchChs.LoadOrStore(key, val)
 	}
-	reqCh := val.(chan Request)
+	tsoQueue := val.(tsoRequestProxyQueue)
 	if !loaded {
+		log.Info("Start new tso proxy dispatcher", zap.String("forwardedHost", req.getForwardedHost()))
 		tsDeadlineCh := make(chan *TSDeadline, 1)
-		go s.dispatch(ctx, tsoProtoFactory, req.getForwardedHost(), req.getClientConn(), reqCh, tsDeadlineCh, doneCh, errCh, tsoPrimaryWatchers...)
-		go WatchTSDeadline(ctx, tsDeadlineCh)
+		dispatcherCtx, ctxCancel := context.WithCancelCause(ctx)
+		tsoQueue.ctx = dispatcherCtx
+		tsoQueue.cancel = ctxCancel
+		go s.dispatch(tsoQueue, tsoProtoFactory, req.getForwardedHost(), req.getClientConn(), tsDeadlineCh, tsoPrimaryWatchers...)
+		go WatchTSDeadline(dispatcherCtx, tsDeadlineCh)
 	}
-	reqCh <- req
+	tsoQueue.request <- req
+	return tsoQueue.ctx
 }
 
 func (s *TSODispatcher) dispatch(
-	ctx context.Context,
+	tsoQueue tsoRequestProxyQueue,
 	tsoProtoFactory ProtoFactory,
 	forwardedHost string,
 	clientConn *grpc.ClientConn,
-	tsoRequestCh <-chan Request,
 	tsDeadlineCh chan<- *TSDeadline,
-	doneCh <-chan struct{},
-	errCh chan<- error,
 	tsoPrimaryWatchers ...*etcdutil.LoopWatcher) {
 	defer logutil.LogPanic()
-	dispatcherCtx, ctxCancel := context.WithCancel(ctx)
-	defer ctxCancel()
 	defer s.dispatchChs.Delete(forwardedHost)
 
-	forwardStream, cancel, err := tsoProtoFactory.createForwardStream(ctx, clientConn)
+	forwardStream, cancel, err := tsoProtoFactory.createForwardStream(tsoQueue.ctx, clientConn)
 	if err != nil || forwardStream == nil {
 		log.Error("create tso forwarding stream error",
 			zap.String("forwarded-host", forwardedHost),
 			errs.ZapError(errs.ErrGRPCCreateStream, err))
-		select {
-		case <-dispatcherCtx.Done():
-			return
-		case _, ok := <-doneCh:
-			if !ok {
-				return
-			}
-		case errCh <- err:
-			close(errCh)
-			return
-		}
+		tsoQueue.cancel(err)
+		return
 	}
 	defer cancel()
 
 	requests := make([]Request, maxMergeRequests+1)
 	needUpdateServicePrimaryAddr := len(tsoPrimaryWatchers) > 0 && tsoPrimaryWatchers[0] != nil
 	for {
+		noProxyRequestsTimer := time.NewTimer(TSOProxyStreamIdleTimeout)
```

Comment:
Do we need to create a timer inside the loop? Is it possible to reuse the same timer always?